File: backend/GameApp.Application/Services/GameServices/GameUpdateService.cs
using GameApp.Application.UseCases.GameUseCases;
using GameApp.Domain.Entities;
using GameApp.Domain.Entities.Scenes;
using GameApp.Domain.Enumerates;
using GameApp.Domain.Repositories;
using GameApp.Domain.ValueObjects.Characters;
using GameApp.Domain.ValueObjects.Enemies;

namespace GameApp.Application.Services.GameServices;

public class GameUpdateService : GameUpdateUseCase
{
    private readonly IGameRepository _repo;

    public GameUpdateService(IGameRepository repo) => _repo = repo;

    public async Task<Game?> UpdateGame(Guid id, Character character, int numberScenesToFinish, List<Scene> completedScenes, NothingHappensScene finalScene, List<Scene> listCurrentScenes, List<UserAction> listCurrentUserActions, GameStatus status, Enemy? currentEnemy)
    {
        var game = await _repo.FetchByIdAsync(id);
        if (game is null)
        {
            Console.WriteLine($"Game to update with id {id} not found."); 
            return null;
        }

        Game updatedGame = game.UpdateGame(character, numberScenesToFinish, completedScenes, finalScene, listCurrentScenes, listCurrentUserActions, status, currentEnemy);
        return await _repo.UpdateAsync(id, updatedGame);
    }
}

File: frontend/gameapp-frontend/src/application/services/game-services/game-update-service.ts
import type { IGameUpdateUseCase } from "../../usecases/game-use-cases/game-update-use-case";
import type { IGameRepository } from "../../repositories/igame-repository";
import { Game } from "../../../domain/entities/game";
import type { GameStatus } from "../../../domain/enumerates/game-status";
import type { FinalScene } from "../../../domain/entities/scenes/final-scene";
import type { Character } from "../../../domain/value-objects/characters/character";
import type { Scene } from "../../../domain/entities/scenes/scene";
import type { UserAction } from "../../../domain/enumerates/user-action";
import type { Enemy } from "../../../domain/entities/enemy";
import type { GameDifficulty } from "../../../domain/enumerates/game-difficulty";

export class GameUpdateService implements IGameUpdateUseCase {
    constructor(private readonly gameRepository: IGameRepository) {}

    async updateGame(
        id: string,
        difficulty: GameDifficulty,
        character: Character,
        numberScenesToFinish: number,
        finalScene: FinalScene,
        listCurrentScenes: Scene[],
        listCurrentUserActions: UserAction[],
        completedScenes: Scene[],
        status: GameStatus,
        currentEnemy: Enemy | null
    ): Promise<Game> {
        const updatedGame = new Game(
            difficulty,
            character,
            numberScenesToFinish,
            finalScene,
            listCurrentScenes,
            listCurrentUserActions,
            completedScenes,
            status,
            currentEnemy,
            id
        );

        return await this.gameRepository.update(id, updatedGame);
    }
}
